filebeat/beater/diagnostics.go

```diff
@@ -0,0 +1,217 @@
+		[]string{"^registry", "filebeat", "[[:digit:]]*\\.json$"},
+	}
+
+	for _, lst := range preFilesList {
+		var path string
+		if filepath.Separator == '\\' {
+			path = strings.Join(lst, `\\`)
+		} else {
+			path = filepath.Join(lst...)
+		}
+
+		// Compile the reg exp, if there is an error, stop and return.
+		// There should be no error here as this code is tested in all
+		// supported OSes, however to avoid a code path that leads to a
+		// panic, we cannot use `regexp.MustCompile` and handle the error
+		re, err := regexp.Compile(path)
+		if err != nil {
+			return nil, fmt.Errorf("cannot compile reg exp: %w", err)
+		}
+
+		registryFileRegExps = append(registryFileRegExps, re)
+	}
+
+	return registryFileRegExps, nil
+}
+
+func gzipRegistry() []byte {
+	logger := logp.L().Named("diagnostics")
+	buf := bytes.Buffer{}
+	dataPath := paths.Resolve(paths.Data, "")
+	registryPath := filepath.Join(dataPath, "registry")
+	f, err := os.CreateTemp("", "filebeat-registry-*.tar")
+	if err != nil {
+		logger.Errorw("cannot create temporary registry archive", "error.message", err)
+	}
+	// Close the file, we just need the empty file created to use it later
+	f.Close()
+	defer logger.Debug("finished gziping Filebeat's registry")
+
+	defer func() {
+		if err := os.Remove(f.Name()); err != nil {
+			logp.L().Named("diagnostics").Warnf("cannot remove temporary registry archive '%s': '%s'", f.Name(), err)
+		}
+	}()
+
+	logger.Debugf("temporary file '%s' created", f.Name())
+	if err := tarFolder(logger, registryPath, f.Name()); err != nil {
+		logger.Errorw(fmt.Sprintf("cannot archive Filebeat's registry at '%s'", f.Name()), "error.message", err)
+	}
+
+	if err := gzipFile(logger, f.Name(), &buf); err != nil {
+		logger.Errorw("cannot gzip Filebeat's registry", "error.message", err)
+	}
+
+	// if the final file is too large, skip it
+	if buf.Len() >= 20_000_000 { // 20 Mb
+		logger.Warnf("registry is too large for diagnostics, %dmb bytes > 20mb", buf.Len()/1_000_000)
+		return nil
+	}
+
+	return buf.Bytes()
+}
+
+// gzipFile gzips src writing the compressed data to dst
+func gzipFile(logger *logp.Logger, src string, dst io.Writer) error {
+	reader, err := os.Open(src)
+	if err != nil {
+		return fmt.Errorf("cannot open '%s': '%w'", src, err)
+	}
+	defer reader.Close()
+
+	writer := gzip.NewWriter(dst)
+	defer writer.Close()
+	writer.Name = filepath.Base(src)
+
+	if _, err := io.Copy(writer, reader); err != nil {
+		if err != nil {
+			return fmt.Errorf("cannot gzip file '%s': '%w'", src, err)
+		}
+	}
+
+	return nil
+}
+
+// tarFolder creates a tar archive from the folder src and stores it at dst.
+//
+// dst must be the full path with extension, e.g: /tmp/foo.tar
+// If src is not a folder an error is retruned
+func tarFolder(logger *logp.Logger, src, dst string) error {
+	fullPath, err := filepath.Abs(src)
+	if err != nil {
+		return fmt.Errorf("cannot get full path from '%s': '%w'", src, err)
+	}
+
+	tarFile, err := os.Create(dst)
```

Comment:
This will be empty if there are no matches, can we avoid including an empty .tar file to avoid people looking into it and wondering if it is supposed to be empty?

If the registry directory exists without a log or filestream input, the regexes will never match for agent Filebeat sub-processes that don't run either of those input types so this will be common.

Reply:
As I mentioned on https://github.com/elastic/beats/pull/41795#discussion_r1906096881 the registry folder is always there and contain some data. Also other inputs can use the registry, from the top of my head: `log`, `filestream` and `journald` all use the registry.

The tar file from this "empty" registry is 3.5kb, the gziped is 228 bytes, it's so little I don't see any problems of having it in the diagnostics. It also works as a sort of "snapshot" of the registry, even if empty, which is a helpful information to have.

Reply:
I don't care about the impact of the size, rather the analysis impact of knowing when its worth decompressing and looking at

A lot of inputs don't use the registry at all, my suspicion is people and tooling will waste time looking at the empty ones because there's no other way to know if its useful besides memorizing which inputs use the registry.

Reply:
Spoke to Tiago about this, the registry isn't empty by default even if unused so detecting this is too annoying.